mypyc/genops.py

```diff
@@ -1610,38 +1660,174 @@ def unbox_or_cast(self, src: Value, target_type: RType, line: int) -> Value:
-        that the name of the function is changed to be '__call__', and a 'self' parameter is added
-        to its list of arguments, as it becomes a class method. The name of the newly constructed
-        class is generated using the names of the functions that enclose the given nested function.
+        This takes a FuncDef and its associated namespace and returns a ClassIR to represent the
+        function. Each callable class contains an environment attribute with points to another
+        ClassIR representing the environment class where some of its variables can be accessed.
+        Note that its '__call__' method is not yet implemented, and is implemented in the
+        add_call_to_callable_class function. The name of the class is generated using the names of
+        the functions that enclose the given nested function.
 
         Returns a newly constructed FuncIR associated with the given FuncDef.
         """
-        class_name = '{}_{}_obj'.format(fdef.name(), namespace)
+        name = '{}_{}_obj'.format(fdef.name(), namespace)
+        count = 0
+        while name in self.callable_class_names:
+            name += '_' + str(count)
+        self.callable_class_names.add(name)
+
+        callable_class = ClassIR(name, self.module_name)
+        callable_class.attributes[ENV_ATTR_NAME] = RInstance(self.func_infos[-2].env_class)
+        callable_class.mro = [callable_class]
+        self.func_infos[-1].callable_class = callable_class
+        self.classes.append(callable_class)
+        return callable_class
+
+    def gen_func_for_callable_class(self,
+                                    blocks: List[BasicBlock],
+                                    sig: FuncSignature,
+                                    env: Environment) -> FuncIR:
+        """Generates a '__call__' method for a callable class representing a nested function.
+
+        This takes the blocks, signature, and environment associated with a function definition and
+        uses those to build the '__call__' method of a given callable class, used to represent that
+        function. Note that a 'self' parameter is added to its list of arguments, as the nested
+        function becomes a class method.
+        """
+        callable_class = self.func_infos[-1].callable_class
         sig = FuncSignature((RuntimeArg('self', object_rprimitive),) + sig.args, sig.ret_type)
-        func_ir = FuncIR('__call__', class_name, self.module_name, sig, blocks, env)
-        class_ir = ClassIR(class_name, self.module_name)
-        class_ir.methods['__call__'] = func_ir
-        self.classes.append(class_ir)
-        return func_ir
+        return FuncIR('__call__', callable_class.name, self.module_name, sig, blocks, env)
 
-    def instantiate_function_class(self, fdef: FuncDef, namespace: str) -> Value:
+    def add_call_to_callable_class(self, call_fn: FuncIR) -> ClassIR:
+        """Sets the '__call__' method for a callable class to a given FuncIR."""
```

Comment:
I'd probably just merge this with `gen_func_for_callable_class`

Reply:
I think it's better to have two separate functions in case in the future we want to do something with any particular one of these.